### PR TITLE
Record autoderef adjustments

### PR DESCRIPTION
**Pull request by Veykril**
cc #9475

### CODE REVIEW

File: crates/hir_ty/src/infer/coerce.rs [hunk 1]
```diff
@@ -324,12 +328,45 @@ impl<'a> InferenceContext<'a> {
...
+    pub(super) fn auto_deref_adjust_steps(&self, autoderef: &Autoderef<'_>) -> Vec<Adjustment> {
+        let steps = autoderef.steps();
+        let targets =
+            steps.iter().skip(1).map(|(_, ty)| ty.clone()).chain(iter::once(autoderef.final_ty()));
+        steps
+            .iter()
+            .map(|(kind, _source)| match kind {
+                AutoderefKind::Overloaded => {
+                    // FIXME: mutability
+                    Some(OverloadedDeref(Mutability::Not))
```
Comment (Veykril) [inline crates/hir_ty/src/infer/coerce.rs:363]: any idea how to go about this?

**Comment by Veykril** [comment]
This doesn't record adjustments for method receiver coercions yet as I'm unsure how to incorporate the special array unsizing for it.

File: crates/hir_ty/src/infer/coerce.rs (hunk 1 above)
Comment (flodiebold) [reply crates/hir_ty/src/infer/coerce.rs:363]: I'm a bit confused about what rustc does. [This](https://github.com/rust-lang/rust/blob/a84d1b21aea9863f0fc5f436b4982d145dade646/compiler/rustc_typeck/src/check/autoderef.rs#L54-L57) seems to be the equivalent place unless I'm missing something, which resolves the overloaded deref and then uses the mutability of the return value; but that just calls [this](https://github.com/rust-lang/rust/blob/a84d1b21aea9863f0fc5f436b4982d145dade646/compiler/rustc_typeck/src/check/place_op.rs#L139-L143), which says it only deals with `Deref`. So I would conclude that this is never mutable. But maybe I'm missing something. It seems the derefs only get converted to mutable in a [separate step](https://github.com/rust-lang/rust/blob/a84d1b21aea9863f0fc5f436b4982d145dade646/compiler/rustc_typeck/src/check/place_op.rs#L221-L228), which gets called depending on the place (i.e. if we have `foo.bar = baz;`, the autoderefs for `foo` would get converted). We don't track whether we need a mutable place during inference yet either, that's also something we should start at some point :smile:

File: crates/hir_ty/src/tests/method_resolution.rs [hunk 2]
```diff
@@ -1137,7 +1139,7 @@ fn main() {
 
 #[test]
 fn autoderef_visibility_field() {
-    check_infer(
+    check(
```
Comment (flodiebold) [inline crates/hir_ty/src/tests/method_resolution.rs:1142]: :+1:

**Review by flodiebold — APPROVED** [review]
LGTM

File: crates/hir_ty/src/infer/coerce.rs (hunk 1 above)
Comment (Veykril) [reply crates/hir_ty/src/infer/coerce.rs:363]: Ye I saw the same stuff there and was kinda confused hence my asking, Makes sense that this is changed up in a different step.

**Comment by Veykril** [comment]
bors r=flodiebold

**Pull request merged**